### DISCUSSION TITLE
Ask method to setup High Availability, HA environment for Mirth Connect 3.5.0.8232

### DISCUSSION THREAD

**Discussion by leejr2022** [nextgenhealthcare/connect#5181] [General]
Hi all, I am new to Mirth Connect, I are having some issue when setup a High availability, HA environment for Mirth Connect 3.5.0.8232. Below is the method we setup the environment.
- Mirth apps: We have two independent Mirth servers with different ip address, behind a load balancer doing round-robin.
- Mirth database: We use Percona XtraDB Cluster to do clustered database.

**Problem faced:** 
1. With the clustered db setup, all the tables for mirth apps must have primary key. However, we found out all the tables in Mirth database dont have primary key. It will cause a lot of warning message in mirth logs as below:

> ERROR (com.mirth.connect.connectors.jdbc.DatabaseReceiver:245): Failed to process row retrieved from the database in channel "SRMS periodical updater v3"
> com.mirth.connect.donkey.server.channel.ChannelException: com.mirth.connect.donkey.server.channel.ChannelException: com.mirth.connect.donkey.server.data.DonkeyDaoException: com.mysql.jdbc.exceptions.jdbc4.MySQLTransactionRollbackException: WSREP detected deadlock/conflict and aborted the transaction. Try restarting the transaction

2. Besides that, we also noticed that we cannot use the Data pruner functionality(settings > Data Pruner) to prune the log message in mirth apps. When we run the Data Pruner, it will cause the clustered database service to down. This is because the database tables do not have primary key and the clustered db try to sync the data during the delete message process, clustered db detected there is data inconsistency and result to stop the db services.

I wish to ask anyone have any idea on how to properly setup the HA environment for Mirth connect application? How should we setup the mirth database as clustered db? Very much appreciate for the help.

**Comment by pacmano1** [comment]
The Advanced Clustering plugin, a paid extension from Nextgen, is "required" if both of those Mirth engines are sharing the same mirth database for their message store.  There are other posts here, and on the Mirth forums that discuss that. And/or what you will need to manage on your own if you don't use that extension.
